Do something useful with SUBSCRIBE_DONE (#609)
moq-wg/moq-transport: draft-ietf-moq-transport.md

A subscriber issues a `UNSUBSCRIBE` message to a publisher indicating it is no
longer interested in receiving media for the specified track and Objects
should stop being sent as soon as possible.  The publisher sends a

> SUBSCRIBE_DONE to acknowledge the unsubscribe was successful.

Reviewer: Seems weird to me that it's named `SUBSCRIBE_DONE` for what is defined as `UNSUBSCRIBE_ACK`

Author: It's not just an UNSUBSCRIBE_ACK. It's also used when the SUBSCRIBE reaches its end.

Reviewer: isn't that redundant with object status 0x4/end of track?

Reviewer: TBH, I don't believe it adds much value to ack unsubscribe.  It should be assumed that it was successful as it was transferred over a reliable control stream, which had to be processed by the remote otherwise the control stream would be invalidated.  If the remote doesn't do anything with the unsubscribe (e.g., noop the message) and continues to publish data, the relay should then respond with error.  I don't recommend using protocol violation in this case as that would close a connection causing a potential large impact that spans more than one track. This is a track error and should, IMHO, only affect that track.  With QUIC,  either side can reset a stream regardless of being bidir or unidir, right?  I would suggest that we need a track level error message that is followed by a RESET of the incoming stream. RESET of stream could indicate error code as well.

Out of scope for this PR but keep in mind as it relates, what happens when an app sends end of track or end of group object status and continues to send data that would violate that status???

Author: It is not redundant to end of track, as other objects could be delivered later and out of order, and sometimes the subscribe ends before the track does

## SUBSCRIBE_DONE {#message-subscribe-done}

A publisher sends a `SUBSCRIBE_DONE` message to indicate it will not open any

> additional streams for a subscription. When all streams for the subscription are

Reviewer 2: Or send additional datagrams

Author: Done. #downwithdatagrams

fully closed, each endpoint can destroy its subscription state.

Note that some objects in the subscribed groups might never be delivered,
because a stream was reset, or never opened in the first place, due to the
delivery timeout.

A sender MUST NOT send SUBSCRIBE_DONE until it has closed all streams it will

> ever open for a subscription. After sending SUBSCRIBE_DONE, MoQT can immediately

Reviewer 2: s/MoQT/a sender (or publisher?)/

> destroy subscription state. The QUIC layer might still be resolving the
> closing of the stream on the wire.

Reviewer 2: ```suggestion
destroy subscription state, even though the underlying transport might not have 
finished closing the associated streams.
```

Reviewer 2: Also do we need to say something about datagrams?

> A receiver that receives SUBSCRIBE_DONE SHOULD set a timer of at least 2 seconds

Reviewer 2: Should there be an exception if the SUBSCRIBE_DONE has code UNSUBSCRIBED (eg: the subscriber already expressed disinterest)?

Author: As the text says, the receiver MAY delete early if the application has lost interest in the data.

I don't think we need special language for that.

Reviewer 3: 2 seconds seems arbitrary.  Can we use 3 * PTO or leave it open to implementations?

Author: It is totally arbitrary. The MoQ application is unlikely to have an API for the PTO, but there's no number I'd be unhappy with.

> in case some unopened streams are still inbound due to prioritization or packet

Reviewer 2: streams or datagrams

## SUBSCRIBE_DONE {#message-subscribe-done}

A publisher sends a `SUBSCRIBE_DONE` message to indicate it will not open any

> additional streams or send any more datagrams for a subscription. When all

Reviewer 4: So lets say I have a track with a single group and single sub group and a million objects in the group. (this might be a common patter for deliver software updates or AI models updates. This new text say the DONE would be sent as soon as the subscribe was received. This seems very wrong and will cause weird race conditions when the DONE was sent after the publisher opened the last stream, but the subscriber got the DONE before they new stream had been opened.

Author: That is not what it says at all. A few paragraphs down, it says:

"A sender MUST NOT send SUBSCRIBE_DONE until it has closed all streams it will
ever open, and has no pending datagrams, for a subscription."

Race conditions are inevitable, even with that rule, because we can't guarantee the order in which QUIC will deliver things. That's why there has to be either a timer or a careful accounting of what streams have been opened.

Reviewer 4: So 

> "A sender MUST NOT send SUBSCRIBE_DONE until it has closed all streams it will
> ever open, and has no pending datagrams, for a subscription."

and 

> "A publisher sends a `SUBSCRIBE_DONE` message to indicate it will not open any
> additional streams or send any more datagrams for a subscription."

Seem to contradict each other about when DONE is sent.

Author: I agree that this is confusing and I'll fix it.

What I'm getting at is this: we don't need a FETCH_DONE because it's all on one stream, so the FIN/RESET is adequate to communicate the end of the FETCH. If the subscriber somehow knew for sure how many streams would open for a SUBSCRIBE, then stream FIN/RESETs would be adequate to know when to destroy subscription state.

SUBSCRIBE_DONE is only needed because it doesn't know for sure; the sender has to tell it that it's done opening streams, so once the current crop of streams is closed it's safe to tear down.

Unfortunately, it is quite likely that SUBSCRIBE_DONE outraces the last stream openings. So either there has to be a timer, or there has to be some indication of how many streams there were (which would levy new requirements on QUIC/WebTransport)

> streams for the subscription are fully closed, each endpoint can destroy its

Reviewer 5: by the phrase "fully closed" , do you mean stream fin or does it cover stream reset cases too ?

Author: It covers Reset -- I'm not sure how it could work just on FIN

subscription state.

> Note that some objects in the subscribed groups might never be delivered,

Reviewer 5: s/subscribed groups/subscribed tracks

because a stream was reset, or never opened in the first place, due to the
delivery timeout.

A sender MUST NOT send SUBSCRIBE_DONE until it has closed all streams it will

> ever open, and has no pending datagrams, for a subscription. After sending

Reviewer 3: ```suggestion
ever open, and has no unsent datagrams, for a subscription. After sending
```

Author: IIUC Datagrams can't be pending at the application level. If QUIC can't take them, they just drop on the floor?

Reviewer 4: The quicr relay queues datagram packets at the applications layer and only passing them to QUIC layer when there is space in the CWIN for them. This allows the relay to control the times outs and priorities for object sent with datagram. I think that will deliver a much better user experience than something that just drops them on the floor (thought both implementations are valid spec compliant AFAIC )

> SUBSCRIBE_DONE, the sender  can immediately destroy subscription state, although

Reviewer 3: ```suggestion
SUBSCRIBE_DONE, the sender can immediately destroy subscription state, although
```

stream state may persist until delivery completes.

> A receiver that receives SUBSCRIBE_DONE SHOULD set a timer of at least 2 seconds

Reviewer 3: ```suggestion
A subscriber that receives SUBSCRIBE_DONE SHOULD set a timer of at least 2 seconds
```

> A subscriber that receives SUBSCRIBE_DONE SHOULD set a timer of at least 2 seconds

Reviewer 4: 2 seconds sounds insanely fragile.

Author: Would you like to suggest a different timeout, or one of the other proposals discussed in this thread? I'm open to any of them but was reluctant to impose more requirements on the QUIC/WebTransport stack.

Author: The effect of the timer expiring early is that some objects may be dropped on the floor instead of delivered to the  application.

If the receiver is interested in objects beyond that, it should set a longer timer.

in case some datagrams or unopened streams are still inbound due to
prioritization or packet loss. Once the timer has expired, the receiver destroys

> subscription state once all open streams for the subscription have closed. A

Reviewer 5: "Once the timer has expired, the receiver can cleanup its subscription state" 

I don't think we need to have "once al open streams for the subscriptions have closed".

> subscription state once all open streams for the subscription have closed. Ai
> receiver MAY destroy subscription state earlier, at the cost of potentially not
> delivering some late objects to the application.

Reviewer 3: ```suggestion
subscription state once all open streams for the subscription have closed. A
subscriber MAY discard subscription state earlier, at the cost of potentially not
delivering some late objects to the application.
```

## SUBSCRIBE_DONE {#message-subscribe-done}

A publisher sends a `SUBSCRIBE_DONE` message to indicate it has finished sending
objects for a subscription, is discarding its subscription state, and the
subscriber can discard its state as well.

A sender MUST NOT send SUBSCRIBE_DONE until all the following conditions are met:

1. All DATAGRAMs sent for the subscription have reached their expiration time.

> 2. All data streams have either been reset, or the stream API has provided a

Reviewer: Is stream API clarified? Suggest rewording as that makes an assumption of an API.

Author: How would you reword it?

Reviewer: I believe the context here is the QUIC transport stack/API, correct?  If so, I would recommend "QUIC transport" instead of "stream API"

notification that the FIN has been ACKed.

3. Any data streams closed with RESET_STREAM_AT have received notification from

> the stream API that the reliable size has been acknowledged.

Reviewer: Same regarding using "stream API". It suggests an implementation that this protocol doesn't appear to define.

  Status Code (i),
  Reason Phrase Length (i),
  Reason Phrase (..),
  ContentExists (8),
  [Final Group (i)],

>   [Final Object (i)],

Reviewer 3: Was this removed because it was difficult, or because you think it's useless or ?

Author: As the description above says, I couldn't think of anything the receiver could do with it. This also clears the issue about if it's the final object of the track or the subscription.

Reviewer 5: i tend to agree with @martinduke  .. a receiver knows the last object it received before it cleans up its subscription state and adding more info in subscribe_done is not that helpful.

Author: @fluffy is it your contention that the Final Object field refers to the final object in the track, or the subscription? If the subscription, is it the final one that was delivered, or the final one that was in range (because those are not the same thing).

> I don't see how this works to remove this. If client A does a subscribe of group 10 to 20, will the relay cache that 20 is the end of track ? I just don't see much connection between end of subscription and end of track.

Why would a relay cache an end of track at all in that case? It is perfectly normal for a cache to not know the end of a track.

> I'm even more baffled how, for a group that is not the last group in the track, this allows the client to figure out when it has got all the objects in the group.

Final Object doesn't ensure that. If a stream is never opened by the sender due to delivery timeout, or the track uses DATAGRAM, the receiver has no idea if those Object IDs were dropped or never minted in the first place.

Reviewer 4: Sorry - in the diff I missed you were removing from the subscribe done only not the dataplane. I don't really case what is there

Reviewer 3: I'm fine with removing this.
